src/librustc/ty/context.rs

```diff
@@ -1082,7 +1089,17 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
 
     #[inline(always)]
     pub fn hir(self) -> &'a hir_map::Map<'gcx> {
-        &self.hir_map
+        let value = self.hir_map.load();
+        if unlikely!(value.is_none()) {
```

Comment:
This should be simplified significantly by the use of `Once`

Reply:
`Once` is slow, and the `hir` function is very hot.

Reply:
ok, can you at least fold this into your `AtomicCell` (by passing a closure to some method) so the non-query code stays out of the query?

Since the `AtomicCell` is just used for this, maybe rename it to `AtomicOnce` and hide the `Option` from the use site.

Reply:
`AtomicCell` is a `crossbeam-utils` type. I can't modify it.

> non-query code stays out of the query

Not sure what you mean here. This is just a regular function.

Reply:
I just saw your version in rustc_datastructures, but I realize now that's just for nonparallel rustc.

Reply:
> Not sure what you mean here. This is just a regular function.

well it's like an emulated query. It would be a regular query just for the caching if it weren't so hot or had any arguments, right?

Which reminds me. Doesn't this being a memoization mean we'll not have an edge from queries calling `hir` to the `hir_map` query, except for the first query to call `hir`? Is that problematic?

Reply:
> except for the first query to call `hir`?

It uses `with_ignore` so there won't be any edges even for the first one. Having edges to the real `hir_map` query would be bad, as it is `no_hash` and `eval_always` which makes it effectively always red (and everything which depends on it always execute).

Not adding edges here is fine since the HIR map adds edges when you use it to access things.

Reply:
I ending up using this pattern in a couple more places so I'm adding a `AtomicOnce` type for it.